elyra/util/cos.py

```diff
@@ -28,36 +28,84 @@ class CosClient(LoggingConfigurable):
             self.endpoint = urlparse(config.metadata['cos_endpoint'])
-            self.access_key = config.metadata.get('cos_username')
-            self.secret_key = config.metadata.get('cos_password')
             self.bucket = config.metadata['cos_bucket']
-        else:
-            self.endpoint = urlparse(endpoint)
-            self.access_key = access_key
-            self.secret_key = secret_key
-            self.bucket = bucket
+            if auth_type in ['USER_CREDENTIALS', 'KUBERNETES_SECRET']:
+                # TODO remove temporary validation checks after
+                # metadata can properly enforce constraints
+                if len(config.metadata.get('cos_username', '').strip()) == 0 or\
+                   len(config.metadata.get('cos_password', '').strip()) == 0:
+                    raise RuntimeError('Cannot connect to object storage. '
+                                       f'Authentication provider \'{auth_type}\' requires '
+                                       'a username and password. Update runtime configuration '
+                                       f'\'{config.display_name}\' and try again.')
+                if auth_type == 'USER_CREDENTIALS' and\
+                   len(config.metadata.get('cos_secret', '').strip()) > 0:
+                    raise RuntimeError('Cannot connect to object storage. '
+                                       f'Authentication provider \'{auth_type}\' does not '
+                                       'support Kubernetes secrets. Update runtime configuration '
+                                       f'\'{config.display_name}\' and try again.')
+                if auth_type == 'KUBERNETES_SECRET' and\
+                   len(config.metadata.get('cos_secret', '').strip()) == 0:
+                    raise RuntimeError('Cannot connect to object storage. '
+                                       f'Authentication provider \'{auth_type}\' requires '
+                                       'a Kubernetes secret name. Update runtime configuration '
+                                       f'\'{config.display_name}\' and try again.')
+                cred_provider = providers.StaticProvider(
+                    access_key=config.metadata['cos_username'],
+                    secret_key=config.metadata['cos_password'],
+                )
+            elif auth_type == 'AWS_ENV_VARIABLES':
+                if os.environ.get('AWS_ACCESS_KEY_ID') is None or\
+                   os.environ.get('AWS_SECRET_ACCESS_KEY') is None:
+                    raise RuntimeError('Cannot connect to object storage. '
+                                       f'Authentication provider \'{auth_type}\' requires '
+                                       'environment variables AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY.')
+                cred_provider = providers.EnvAWSProvider()
+            elif auth_type == 'AWS_IAM_ROLES_FOR_SERVICE_ACCOUNTS':
+                if os.environ.get('AWS_ROLE_ARN') is None or\
```

Comment:
Ditto here - could be part of a save hook.

Reply:
I did consider validating the existence of those environment variables but ultimately decided against it:
- it's a valid scenario for the variables to not exist until they are needed (an analogy here would be if we were to prevent saving of a KFP runtime configuration if the specified kubeflow server cannot be "pinged" - e.g. because one forgot to connect to a VPN)
- the metadata service doesn't seem to support yet warnings that alert the user to the fact that something in the configuration might be wrong but is not treated as a fatal error (if there is please let me know!)

Reply:
No worries.  Just had some responses to your points.

> it's a valid scenario for the variables to not exist until they are needed

Since these env vars are in the environment of the Elyra process, how are they going to come into existence _after_ the process has started?  Although Python does allow a running process's environment to be modified (Java does not), we don't provide any mechanisms for setting these variables after startup (nor should we IMO).

> the metadata service doesn't seem to support yet warnings that alert the user to the fact that something in the configuration might be wrong but is not treated as a fatal error 

Correct, this is by design.  The metadata service is agnostic to the context of its instances other than what can be expressed in JSON schema.  However, schema validation is binary (success or failure), so warnings are not something that will be produced by the Metadata Service itself.  However, this is one of the reasons we support custom metadata classes since they DO have contextual awareness and could produce similar "warnings" (which appear to be `RuntimeError`s) relative to a given instance.

I'm curious how we handle this with anything other than an exception from the server since the request has to fail.  What is your notion of a _fatal error_ vs. a _warning_ in this particular instance (where an `auth_type` doesn't have its required properties)?

Reply:
> Since these env vars are in the environment of the Elyra process, how are they going to come into existence after the process has started?

I was thinking of container-based deployments, where metadata is persisted across invocations, yet the environment settings can change.

> What is your notion of a fatal error vs. a warning in this particular instance (where an auth_type doesn't have its required properties)?

I was thinking that a warning could alert the user that some action needs to be taken for this configuration to be ready for use. In this specific scenario a message could indicate that environment variables XYZ1 and XYZ2 are required but not set. Since we currently don't explicitly expose prerequisites in the UI ("requires environment variables 'XYZ1' and 'XYZ2' to be set" ) users probably won't have them set in their Elyra environment the first time they try to use an RTC with this authentication type.

Reply:
> I was thinking of container-based deployments, where metadata is persisted across invocations, yet the environment settings can change.

Ok, but since the code that is checking for the environment variable _is_ the server _process_, if the envs are not present, the container would require a redeployment (where the variables are set via the container's CLI or helm chart, etc.) - just like you'd need to restart the process (with the envs present) in non-container configurations.

> I was thinking that a warning could alert the user that some action needs to be taken for this configuration to be ready for use.  ...

I see.  The request would still fail, but the message returned with that failure would be phrased as a warning - am I understanding this correctly?

Reply:
> > I was thinking of container-based deployments, where metadata is persisted across invocations, yet the environment settings can change.
> 
> Ok, but since the code that is checking for the environment variable _is_ the server _process_, if the envs are not present, the container would require a redeployment (where the variables are set via the container's CLI or helm chart, etc.) - just like you'd need to restart the process (with the envs present) in non-container configurations.

If you mean by 'redeployment' restart, then I believe the answer is yes. One can, for example, pod defaults that would take effect when a container comes up. 


> > I was thinking that a warning could alert the user that some action needs to be taken for this configuration to be ready for use.  ...
> 
> I see. The request would still fail, but the message returned with that failure would be phrased as a warning - am I understanding this correctly?
Yes!